### PR TITLE
Fix algorithmic complexity of on-demand scheduler with regards to number of cores.

### PR DESCRIPTION
**Pull request by eskimor**
We witnessed really poor performance on Rococo, where we ended up with 50 on-demand cores. This was due to the fact that for each core the full queue was processed. With this change full queue processing will happen way less often (most of the time complexity is O(1) or O(log(n))) and if it happens then only for one core (in expectation).

TODO:

- [ ] Implement
- [ ] Basic tests
- [ ] Add more tests (see todos)
- [ ] Run benchmark to confirm better performance
- [ ] Write migrations
- [ ] Bump scale-info version and remove patch in Cargo.toml

Optional: Performance can be improved even more, if we called `pop_assignment_for_core()`, before calling `report_processed` (Avoid needless affinity drops). The effect gets smaller the larger the claim queue and I would only go for it, if it does not add complexity to the scheduler.

### CODE REVIEW

File: polkadot/runtime/parachains/src/assigner_on_demand/mod.rs [hunk 1]
```diff
@@ -1,4 +1,4 @@
-// Copyright (C) Parity Technologies (UK) Ltd.
+// Copyright (C) Parit (UK) Ltd.
```
Comment (ordian) [inline polkadot/runtime/parachains/src/assigner_on_demand/mod.rs:1]: 👀

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: check-tracing
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5107984

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 1/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5107966

File: polkadot/runtime/parachains/src/assigner_on_demand/mod.rs (hunk 1 above)
Comment (eskimor) [reply polkadot/runtime/parachains/src/assigner_on_demand/mod.rs:1]: LOL. Thanks for catching this. Must have been a rogue vim key binding :see_no_evil:

File: polkadot/runtime/parachains/src/assigner_on_demand/mod.rs [hunk 2]
```diff
@@ -429,172 +664,104 @@ where
...
-			},
+	/// if affinity dropped to 0, moving entries back to `FreeEntries`.
+	fn decrease_affinity_update_queue(para_id: ParaId, core_index: CoreIndex) {
+		let affinity = Pallet::<T>::decrease_affinity(para_id, core_index);
+		#[cfg(not(test))]
+		debug_assert_ne!(
+			affinity, None,
+			"Decreased affinity for a para that has not been served on a core?"
+		);
+		if affinity != Some(0) {
```
Comment (ordian) [inline polkadot/runtime/parachains/src/assigner_on_demand/mod.rs:702]: should this be `==`?

File: polkadot/runtime/parachains/src/assigner_on_demand/mod.rs (hunk 1 above)
Comment (ordian) [reply polkadot/runtime/parachains/src/assigner_on_demand/mod.rs:1]: I thought you might know something I don't 😅

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: check-tracing
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5109260

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: check-tracing
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5110311

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5110276

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-int
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5110310

**Comment by ordian** [comment]
New profile looks good 🚀 
new-profile

File: polkadot/runtime/parachains/src/assigner_on_demand/mod.rs [hunk 3]
```diff
@@ -297,10 +434,83 @@ pub mod pallet {
...
 
+// Internal functions and interface to scheduler/wrapping assignment provider.
 impl<T: Config> Pallet<T>
 where
 	BalanceOf<T>: FixedPointOperand,
 {
+	/// Take the next queued entry that is available for a given core index.
+	/// Invalidates and removes orders with a `para_id` that is not `ParaLifecycle::Parathread`
+	/// but only in [0..P] range slice of the order queue, where P is the element that is
+	/// removed from the order queue.
```
Comment (ordian) [inline polkadot/runtime/parachains/src/assigner_on_demand/mod.rs:445]: the bit about `ParaLifecycle::Parathread` seems to be outdated in this PR - don't see any checks for paralifecycle here

File: polkadot/runtime/parachains/src/assigner_on_demand/migration.rs [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+				Pallet::<T>::update_spot_traffic(&config, &mut queue_status);
+
+				let v0_queue = v0::OnDemandQueue::<T>::take();
+				// Process the v0 queue into v1.
+				v0_queue.into_iter().for_each(|enqueued_order| {
+					// Readding the old orders will use the new systems.
+					Pallet::<T>::add_on_demand_order(
+						queue_status,
+						enqueued_order.para_id,
+						QueuePushDirection::Front,
```
Comment (eskimor) [inline polkadot/runtime/parachains/src/assigner_on_demand/migration.rs:70]: Aren't we reversing the order here?

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-int
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5464308

File: polkadot/runtime/parachains/src/assigner_on_demand/migration.rs [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+				v0::OnDemandQueue::<T>::get().is_empty(),
+				"OnDemandQueue should be empty after the migration"
+			);
+
+			let expected_len = u32::decode(&mut &state[..]).unwrap();
+			let n_affinity_entries = AffinityEntries::<T>::iter()
+				.map(|(_index, heap)| heap.len() as u32)
+				.fold(0, |acc, x| acc + x);
+			let n_free_entries = FreeEntries::<T>::get().len() as u32;
+			let n_orders = n_affinity_entries + n_free_entries;
```
Comment (eskimor) [inline polkadot/runtime/parachains/src/assigner_on_demand/migration.rs:116]: Why not use [size](https://github.com/paritytech/polkadot-sdk/blob/ce7c8e846f0e5adfa33c488849209d0fa3ac6079/polkadot/runtime/parachains/src/assigner_on_demand/mod.rs#L137)? Should be good enough, if we have unit tests for size.

Comment (eskimor) [reply polkadot/runtime/parachains/src/assigner_on_demand/migration.rs:116]: But if we want to do a full check, we could also compare this to `size` in addition.

File: polkadot/runtime/parachains/src/assigner_on_demand/tests.rs [hunk 6]
```diff
@@ -211,6 +214,42 @@ fn spot_traffic_decreases_over_time() {
...
+		OnDemandAssigner::set_queue_status(QueueStatusType {
+			traffic: FixedU128::from_u32(10),
+			..Default::default()
+		});
+
+		assert_eq!(OnDemandAssigner::get_queue_status().traffic, FixedU128::from_u32(10));
+
+		// Run to block 101 and ensure that the traffic decreases.
+		run_to_block(101, |n| if n == 100 { Some(Default::default()) } else { None });
+		assert_eq!(OnDemandAssigner::get_queue_status().traffic, FixedU128::from_float(2.49996875));
```
Comment (eskimor) [inline polkadot/runtime/parachains/src/assigner_on_demand/tests.rs:242]: A more robust check would be to just check that the value got smaller.

File: polkadot/runtime/parachains/src/assigner_on_demand/tests.rs [hunk 7]
```diff
@@ -497,3 +536,127 @@ fn affinity_changes_work() {
...
+	// the overflow will cause an opposite comparison to what would be expected.
+	let mut max_num = u32::MAX - ON_DEMAND_MAX_QUEUE_MAX_SIZE;
+	// 0 < some large number.
+	assert_eq!(QueueIndex(0).cmp(&QueueIndex(max_num + 1)), Ordering::Less);
+}
+
+#[test]
+fn queue_index_ordering_works() {
+	// The largest number accepted in in the queue index.
+	let mut max_num = ON_DEMAND_MAX_QUEUE_MAX_SIZE;
```
Comment (eskimor) [inline polkadot/runtime/parachains/src/assigner_on_demand/tests.rs:613]: The index is actually allowed to be larger, just the size of the queue must never exceed this value.

File: polkadot/runtime/parachains/src/assigner_on_demand/tests.rs [hunk 8]
```diff
@@ -497,3 +536,127 @@ fn affinity_changes_work() {
...
+	// Ordering with an element outside of the bounds of the max size also works.
+	v = vec![max_num + 2, 0, 6, 2, 1, 5, 4];
+	v.sort_by_key(|&num| QueueIndex(num));
+	assert_eq!(v, vec![0, 1, 2, 4, 5, 6, max_num + 2]);
+
+	// Numbers way above the max size are unsound and will fail to be sorted.
+	// Assume all numbers above the max size to be unsound.
+	v = vec![u32::MAX, 6, 2, 1, 5, 4];
+	v.sort_by_key(|&num| QueueIndex(num));
+	assert_eq!(v, vec![u32::MAX, 1, 2, 4, 5, 6]);
```
Comment (eskimor) [inline polkadot/runtime/parachains/src/assigner_on_demand/tests.rs:644]: This is actually "sound" and what we expect with the overflow. They are sorted and exactly how we want them to be sorted.

If the queue index is overflowing, then this sort is exactly what we want. Consider the following queue:

begin: u32::MAX - 1
end: 5

Entries in the queue would be: [u32::MAX-1, u32::MAX, 0, 1, 2, 3, 4] ... exactly what we want.

**Review by eskimor — COMMENTED** [review]
Nice work, thanks @antonva !

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5531930

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 2/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5531949

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 3/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5531950

File: polkadot/runtime/parachains/src/assigner_on_demand/migration.rs [hunk 9]
```diff
@@ -67,7 +67,7 @@ mod v1 {
 					Pallet::<T>::add_on_demand_order(
 						queue_status,
 						enqueued_order.para_id,
-						QueuePushDirection::Front,
+						QueuePushDirection::Back,
```
Comment (eskimor) [inline polkadot/runtime/parachains/src/assigner_on_demand/migration.rs:70]: Maybe it would be a good idea to add a [migration test](https://github.com/paritytech/polkadot-sdk/pull/3479/files#diff-8247e57109bbaa6dd0f7271784ad785d3240a2e72c081ddf7c658c08deedde85R143), to make sure we are preserving the order correctly.

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 3/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5535483

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 2/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5535482

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5535463

File: polkadot/runtime/parachains/src/assigner_on_demand/migration.rs [hunk 10]
```diff
@@ -136,3 +136,44 @@ pub type MigrateV0ToV1<T> = VersionedMigration<
 	Pallet<T>,
 	<T as frame_system::Config>::DbWeight,
 >;
+
+#[cfg(test)]
+mod tests {
```
Comment (eskimor) [inline polkadot/runtime/parachains/src/assigner_on_demand/migration.rs:141]: Nice! :-)

File: polkadot/runtime/parachains/src/assigner_on_demand/mod.rs [hunk 11]
```diff
@@ -73,17 +87,118 @@ impl WeightInfo for TestWeightInfo {
...
+///
+///  - insert: O(1) for placing an order, O(log(N)) for push backs.
+///  - pop_assignment_for_core: O(log(N)), O(N) worst case: Can only happen for one core, next core
+///  is already less work.
+///  - report_processed & push back: If affinity dropped to 0, then O(N) in the worst case. Again
+///  this divides per core.
+///
+///  Reads still exist, also improved slightly, but worst case we fetch all entries.
+///
+///  TODO: Add benchmark results.
```
Comment (ordian) [inline polkadot/runtime/parachains/src/assigner_on_demand/mod.rs:112]: still missing

File: polkadot/runtime/parachains/src/assigner_on_demand/mod.rs [hunk 12]
```diff
@@ -104,15 +218,86 @@ pub enum SpotTrafficCalculationErr {
...
+		Self { idx, para_id }
+	}
+}
+
+impl PartialOrd for EnqueuedOrder {
+	fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+		match other.idx.partial_cmp(&self.idx) {
+			None => return None,
+			Some(Ordering::Equal) => other.para_id.partial_cmp(&self.para_id),
+			Some(o) => Some(o),
```
Comment (ordian) [inline polkadot/runtime/parachains/src/assigner_on_demand/mod.rs:290]:
```suggestion
			Some(Ordering::Equal) => other.para_id.partial_cmp(&self.para_id),
			o => o,
```

**Review by ordian — COMMENTED** [review]
Anything missing apart from running the benches for testnets?

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5560526

File: polkadot/runtime/parachains/src/assigner_on_demand/mod.rs [hunk 13]
```diff
@@ -430,175 +685,140 @@ where
...
-				return config.scheduler_params.on_demand_queue_max_size
-			},
+	/// if affinity dropped to 0, moving entries back to `FreeEntries`.
+	fn decrease_affinity_update_queue(para_id: ParaId, core_index: CoreIndex) {
+		let affinity = Pallet::<T>::decrease_affinity(para_id, core_index);
+		#[cfg(not(test))]
+		debug_assert_ne!(
+			affinity, None,
+			"Decreased affinity for a para that has not been served on a core?"
+		);
```
Comment (ordian) [inline polkadot/runtime/parachains/src/assigner_on_demand/mod.rs:722]: first time seeing debug assert for non test environment. is the intention to run this assert on testnets and benchmarks only?

**Review by ordian — APPROVED** [review]
Approving modulo nits, but I'm not deeply familiar with the underlying logic of the on_demand assigner

File: polkadot/runtime/parachains/src/assigner_on_demand/mod.rs [hunk 14]
```diff
@@ -73,17 +87,116 @@ impl WeightInfo for TestWeightInfo {
...
 }
 
+/// Meta data for full queue.
+///
+/// This includes elements with affinity and free entries.
+///
+/// The actual queue is implemented via multiple priority queues. One for each core, for entries
+/// which currently have a core affinity and one free queue, with entries without any affinity yet.
+///
+/// The design aims to have most queue accessess be O(1) or O(log(N)). Absolute worst case is O(N).
```
[the remark this answers is not in the document]
Comment (eskimor) [reply polkadot/runtime/parachains/src/assigner_on_demand/mod.rs:97]: You mean, because the data still needs to be fetched? Yes indeed, I considered this, but decided to not make it part of the complexity analysis, because the fetching of data is unavoidable and not really part of the algorithm. I was also hoping that the batch data fetching is fast. In any case, we are also usually fetching less data than before. I was tempted to optimize not accessing the free list, in those cases where we don't need data from there, by keeping track of its status, but decided the complexity is not worth it.

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5581577

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 3/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5581597

File: polkadot/runtime/westend/src/weights/runtime_parachains_assigner_on_demand.rs [hunk 15]
```diff
@@ -48,44 +48,44 @@ use core::marker::PhantomData;
...
-		//  Estimated: `3762 + s * (4 ±0)`
-		// Minimum execution time: 33_522_000 picoseconds.
-		Weight::from_parts(35_436_835, 0)
-			.saturating_add(Weight::from_parts(0, 3762))
-			// Standard Error: 129
-			.saturating_add(Weight::from_parts(14_041, 0).saturating_mul(s.into()))
+		//  Measured:  `218 + s * (8 ±0)`
+		//  Estimated: `3681 + s * (8 ±0)`
+		// Minimum execution time: 21_396_000 picoseconds.
+		Weight::from_parts(20_585_695, 0)
```
Comment (eskimor) [inline polkadot/runtime/westend/src/weights/runtime_parachains_assigner_on_demand.rs:63]: :partying_face:

File: polkadot/runtime/parachains/src/assigner_on_demand/mod.rs (hunk 13 above)
Comment (eskimor) [reply polkadot/runtime/parachains/src/assigner_on_demand/mod.rs:722]: I guess it breaks some edge case tests?
